lib/framedivider.rb

```diff
@@ -390,46 +422,284 @@ def opening(model, id)
+  ptz[:D][:from_next].normalize! if four
+  ptz[:D][:n_prev_unto_next_pl].normalize! if four
+  ptz[:D][:p] = ptz[:D][:pt] + (ptz[:D][:n_prev_unto_next_pl] * width) + (ptz[:D][:from_next] * width * Math.tan(ptz[:D][:angle]/2)) if four
 
-  # Re-convert Topolys 3D points into OpenStudio 3D points for fitting test.
-  vec = OpenStudio::Point3dVector.new
-  vec << OpenStudio::Point3d.new(pts[:A][:p].x, pts[:A][:p].y, pts[:A][:p].z)
-  vec << OpenStudio::Point3d.new(pts[:B][:p].x, pts[:B][:p].y, pts[:B][:p].z)
-  vec << OpenStudio::Point3d.new(pts[:C][:p].x, pts[:C][:p].y, pts[:C][:p].z)
-  vec << OpenStudio::Point3d.new(pts[:D][:p].x, pts[:D][:p].y, pts[:D][:p].z) if four
-
-  return u, area, points unless fit?(model, id, vec)
-
-  tr1 = []
-  tr1 << pts[:A][:p]
-  tr1 << pts[:B][:p]
-  tr1 << pts[:C][:p]
-  area1 = areaHeron(tr1)
-
-  area2 = 0
-  if four
-    tr2 = []
-    tr2 << pts[:A][:p]
-    tr2 << pts[:C][:p]
-    tr2 << pts[:D][:p]
-    area2 = areaHeron(tr2)
-  end
-  area = area1 + area2
-
-  return u, area, vec
+  ptz
+end
+
+##
+# Fetch a surface's subsurface opening areas & vertices.
+#
+# @param [OpenStudio::Model::Model] model An OpenStudio model
+# @param [OpenStudio::Model::Surface] surface An OpenStudio surface
+#
+# @return [Hash] Returns surface with key attributes, including openings.
+def openings(model, surface)
+  surf = {}
+
+  unless model && model.is_a?(OpenStudio::Model::Model)
+    msg = "Can't find OpenStudio model for 'openings' - skipping"
+    TBD.log(TBD::DEBUG, msg)
+    return nil
+  end
+  unless surface && surface.is_a?(OpenStudio::Model::Surface)
+    msg = "Can't find OpenStudio surface for 'openings' - skipping"
+    TBD.log(TBD::DEBUG, msg)
+    return nil
+  end
+
+  identifier = surface.nameString
+
+  if surface.space.empty?
+    msg = "Can't find space holding '#{identifier}' - skipping"
+    TBD.log(TBD::ERROR, msg)
+    return nil
+  end
+  surf[:space] = surface.space.get
+
+  a = surf[:space].spaceType.empty?
+  surf[:stype] = surf[:space].spaceType.get unless a
+  a = surf[:space].buildingStory.empty?
+  surf[:story] = surf[:space].buildingStory.get unless a
+
+  # Site-specific (or absolute, or true) surface normal.
+  t, r = transforms(model, surf[:space])
+  unless t && r
+    msg = "Can't process '#{identifier}' space transformation"
+    TBD.log(TBD::FATAL, msg)
+    return nil
+  end
+
+  n = trueNormal(surface, r)
+  unless n
+    msg = "Can't process '#{identifier}' true normal"
+    TBD.log(TBD::FATAL, msg)
+    return nil
+  end
+  surf[:n] = n
+
+  surf[:gross] = surface.grossArea
+  fd = false
+  subs = {}
+
+  # Option A:
+  # subsurfaces = model.getSubSurfaces.sort_by{ |s| s.nameString }
+  # subsurfaces.each do |s|
+
+  # Option B:
+  # surface.subSurfaces.each do |s|
+
+  # Option C:
+  # model.getModelObjects.each do |s|
+    # next if s.to_SubSurface.empty?
+    # s = s.to_SubSurface.get
+
+  # Option D:
+  subsurfaces = surface.subSurfaces.sort_by { |s| s.nameString }
```

Comment:
I'd prefer sorting by area so the results don't change if you rename a surface

Reply:
Subsurfaces like skylights usually have the same area (+50 skylights in the warehouse benchmark model). I get your point, but we can at least depend on the uniqueness of the name (not the area).

Reply:
Ah I see your point, you can also sort by both area and name:

surface.subSurfaces.sort_by { |s| "#{s.grossArea}-#{s.nameString}" }

Either way will give you the same result on the same model (which is good).  The thing I am concerned about is getting different results when you change the model in minor ways (e.g. renaming a subsurface).  Any type of sorting is an improvement over no sorting.